text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,810 @@
+you could implement your own `Rc` type like this:
+
+```rust
+#[derive(SmartPointer)]
+pub struct Rc<T: ?Sized> {
+    inner: NonNull<RcInner<T>>,
+}
+
+struct RcInner<T: ?Sized> {
+    refcount: usize,
+    value: T,
+}
+
+impl<T: ?Sized> Deref for Rc<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        let ptr = self.inner.as_ptr();
+        unsafe { &*ptr.value }
+    }
+}
+
+impl<T> Rc<T> {
+    pub fn new(value: T) -> Self {
+        let inner = Box::new(RcInner {
+            refcount: 1,
+            value,
+        });
+        Self {
+            inner: NonNull::from(Box::leak(inner)),
+        }
+    }
+}
+
+impl<T: ?Sized> Clone for Rc<T> {
+    fn clone(&self) -> Self {
+        unsafe { (*self.inner.as_ptr()).refcount += 1 };
+        Self { inner: self.inner }
+    }
+}
+
+impl<T: ?Sized> Drop for Rc<T> {
+    fn drop(&mut self) {
+        let ptr = self.inner.as_ptr();
+        unsafe { (*ptr).refcount -= 1 };
+        if unsafe { (*ptr).refcount } == 0 {
+            drop(unsafe { Box::from_raw(ptr) });
+        }
+    }
+}
+```
+In this example, `#[derive(SmartPointer)]` makes it possible to use `Rc<dyn
+MyTrait>`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The derive macro will expand into two trait implementations,
+[`core::ops::CoerceUnsized`] to enable unsizing coercions and
+[`core::ops::DispatchFromDyn`] for dynamic dispatch. This expansion will be
+adapted in the future if the underlying mechanisms for unsizing coercions and
+dynamically dispatched receivers changes.
+
+As mentioned in the [rationale][why-only-macro] section, this RFC only proposes
+to stabilize the derive macro. The underlying traits used by its expansion will
+remain unstable for now.
+
+## Input Requirements
+[input-requirements]: #input-requirements
+
+The macro sets the following requirements on its input:
+
+1. The definition must be a struct.
+2. The struct must have at least one type parameter. If multiple type
+   parameters are present, exactly one of them has to be annotated with the
+   `#[pointee]` derive helper attribute.
+3. The struct must not be `#[repr(packed)]` or `#[repr(C)]`.
+4. Other than one-aligned, zero-sized fields, the struct must have exactly one
+   field.
+5. Assume that `T` is a type that can be unsized to `U`, and let `FT` and `FU`
+   be the type of the struct's field when the pointee is equal to `T` and `U`
+   respectively. If the struct's trait bounds are satisfied for both `T` and
+   `U`, then it must be possible to convert `FT` to `FU` using an unsizing
+   coercion.
+
+(Adapted from the docs for [`DispatchFromDyn`].)
+
+Point 1 and 2 are verified syntactically by the derive macro, whereas 3, 4 and 5
+are verified semantically by the compiler when checking the generated
+[`DispatchFromDyn`] implementation as it does today.
+
+The `#[pointee]` attribute may also be written as `#[smart_pointer::pointee]`.
```

Comment:
I understand the idea of an informal snake_case/CamelCase correspondence, but I still want to identify something more disciplined here than just than informal correspondence.

E.g. can we treat the underlying idea here that the `SmartPointer` "thing" being derived is a lang item coming from `core::smart_pointer::SmartPointer`?

Alternative, could the elaborated `#[pointee]` attribute be written as `#[SmartPointer::pointee]`? 

I basically want to push back very lightly against grabbing *both* `SmartPointer` and `smart_pointer` without having some relationship between them as items in the modules provided by core.